features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/aes_alt.c

```diff
@@ -0,0 +1,745 @@
+                    unsigned char iv[16],
+                    const unsigned char *input,
+                    unsigned char *output )
+{
+    unsigned char temp[CY_CRYPTO_AES_BLOCK_SIZE];
+
+    int ret = 0;
+    cy_en_crypto_status_t status;
+
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+    AES_VALIDATE_RET( iv != NULL );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+
+    if( length % CY_CRYPTO_AES_BLOCK_SIZE )
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+
+    AES_VALIDATE_RET( ctx->aes_state.buffers != NULL);
+
+    if( mode == MBEDTLS_AES_DECRYPT )
+    {
+        while(( length > 0 ) && (ret == 0))
+        {
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, temp, input, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_DECRYPT, output, input, &ctx->aes_state);
+
+            Cy_Crypto_Core_MemXor(ctx->obj.base, output, output, iv, CY_CRYPTO_AES_BLOCK_SIZE);
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, iv, temp, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            input  += CY_CRYPTO_AES_BLOCK_SIZE;
+            output += CY_CRYPTO_AES_BLOCK_SIZE;
+            length -= CY_CRYPTO_AES_BLOCK_SIZE;
+
+            if (CY_CRYPTO_SUCCESS != status)
+            {
+                ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+            }
+        }
+    }
+    else
+    {
+        while(( length > 0 ) && (ret == 0))
+        {
+            Cy_Crypto_Core_MemXor(ctx->obj.base, output, input, iv, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_ENCRYPT, output, output, &ctx->aes_state);
+
+            Cy_Crypto_Core_MemCpy(ctx->obj.base, iv, output, CY_CRYPTO_AES_BLOCK_SIZE);
+
+            input  += CY_CRYPTO_AES_BLOCK_SIZE;
+            output += CY_CRYPTO_AES_BLOCK_SIZE;
+            length -= CY_CRYPTO_AES_BLOCK_SIZE;
+
+            if (CY_CRYPTO_SUCCESS != status)
+            {
+                ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+            }
+        }
+    }
+
+    return( ret );
+}
+
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+
+/* Endianess with 64 bits values */
+#ifndef GET_UINT64_LE
+#define GET_UINT64_LE(n,b,i)                            \
+{                                                       \
+    (n) = ( (uint64_t) (b)[(i) + 7] << 56 )             \
+        | ( (uint64_t) (b)[(i) + 6] << 48 )             \
+        | ( (uint64_t) (b)[(i) + 5] << 40 )             \
+        | ( (uint64_t) (b)[(i) + 4] << 32 )             \
+        | ( (uint64_t) (b)[(i) + 3] << 24 )             \
+        | ( (uint64_t) (b)[(i) + 2] << 16 )             \
+        | ( (uint64_t) (b)[(i) + 1] <<  8 )             \
+        | ( (uint64_t) (b)[(i)    ]       );            \
+}
+#endif
+
+#ifndef PUT_UINT64_LE
+#define PUT_UINT64_LE(n,b,i)                            \
```

Comment:
Two issues:
1. If any of the input parameters have a side effect then this macro will not work as intended. 
2. What happens if 'n' is signed and negative? C standard says undefined; could be sign extended.

Reply:
This code was captured to support XTS cipher mode from aes.c. Please look into line 1129.